File: C_Programming/Practise_Programs/PP_132.c
//////////////////////////////////////////////////////////////////
//
//  File Name         : PP_132
//
//  Description       : This program demonstrates reverse traversal
//                      of a string using a character pointer and
//                      displays the string from different positions.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 16/08/2026
//
//  Time Complexity   : O(n^2)
//  Space Complexity  : O(n)
//
//////////////////////////////////////////////////////////////////

#include<stdio.h>

//////////////////////////////////////////////////////////////////
//
//  Function Name     : ReverseDisplay()
//
//  Description       : It is used to traverse the string in reverse
//                      order using a character pointer and display
//                      the string from different positions.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 16/08/2026
//
//////////////////////////////////////////////////////////////////

void ReverseDisplay(char *str)
{
    char *start  = NULL;
    
    start = str;

    while(*str != '\0')
    {
        str++;
    }

    str--;
    
    while(start <= str)
    {
        printf("%s\n", str);
        str--;
    }

    printf("\n");
}

int main()
{
    char Arr[50] = {'\0'};

    printf("Enter String : \n");
    scanf("%[^'\n]s", Arr);

    ReverseDisplay(Arr);

    return 0;
}

//////////////////////////////////////////////////////////////////
//
//  Sample Input :
//      Ganesh
//
//  Sample Output :
//
//      Enter String :
//      h
//      h
//      h
//      h
//      h
//      h
//
//////////////////////////////////////////////////////////////////

File: Java_Programming/Practise_Programs/PP_132.java
//////////////////////////////////////////////////////////////////
//
//  File Name         : PP_132
//
//  Description       : This program accepts a string from the user
//                      and converts lowercase characters into
//                      uppercase characters using ASCII values.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 30/08/2026
//
//  Time Complexity   : O(n)
//  Space Complexity  : O(n)
//
//////////////////////////////////////////////////////////////////

import java.util.Scanner;

class StringX
{
    //////////////////////////////////////////////////////////////////
    //
    //  Method Name      : toUpperX()
    //
    //  Description      : Converts the given string into a
    //                      character array and subtracts 32 from
    //                      every character to convert it into
    //                      uppercase.
    //
    //  Issue            : The program subtracts 32 from all
    //                      characters, including uppercase letters,
    //                      digits, spaces and special symbols.
    //
    //////////////////////////////////////////////////////////////////

    public String toUpperX(String str)
    {
        int i = 0;

        char Arr[] = str.toCharArray();

        for(i = 0; i < Arr.length; i++)
        {
            Arr[i] = (char) (Arr[i] - 32);     // ISSUE
        }

        return new String(Arr);                
    }
}

class PP_132
{
    public static void main(String A[])
    {
        Scanner sobj = new Scanner(System.in);
        String data = null;

        StringX strobj = new StringX();
        String sRet = null;

        System.out.println("Enter String : ");
        data = sobj.nextLine();

        sRet = strobj.toUpperX(data);

        System.out.println("Updated string is : " + sRet);

        sobj.close();
    }
}

//////////////////////////////////////////////////////////////////
//
//  Sample Input :
//      Ganesh 123
//
//  Sample Output :
//
//      Incorrect Output
//
//      Reason:
//      The program subtracts 32 from every character instead of
//      checking whether the character is between 'a' and 'z'.
//
//////////////////////////////////////////////////////////////////